d/src/google/protobuf/struct_.d

```diff
@@ -0,0 +1,218 @@
+
+        final switch (kindCase)
+        {
+        case KindCase.kindNotSet:
+            return JSONValue(null);
+        case KindCase.nullValue:
+            return JSONValue(null);
+        case KindCase.numberValue:
+            return JSONValue(numberValue);
+        case KindCase.stringValue:
+            return JSONValue(stringValue);
+        case KindCase.boolValue:
+            return JSONValue(boolValue);
+        case KindCase.structValue:
+            return JSONValue(structValue.fields.byKeyValue.map!(a => tuple(a.key, a.value.toJSONValue)).assocArray);
+        case KindCase.listValue:
+            return JSONValue(listValue.values.map!(a => a.toJSONValue).array);
+        }
+    }
+
+    auto fromJSONValue()(JSONValue jsonValue)
+    {
+        import std.array : array, assocArray;
+        import std.algorithm : map;
+        import std.json : JSON_TYPE;
+        import std.typecons : tuple;
+
+        switch (jsonValue.type)
+        {
+        case JSON_TYPE.NULL:
+            nullValue = NullValue.NULL_VALUE;
+            break;
+        case JSON_TYPE.STRING:
+            stringValue = jsonValue.str;
+            break;
+        case JSON_TYPE.INTEGER:
+            numberValue = jsonValue.integer;
+            break;
+        case JSON_TYPE.UINTEGER:
+            numberValue = jsonValue.uinteger;
+            break;
+        case JSON_TYPE.FLOAT:
+            numberValue = jsonValue.floating;
+            break;
+        case JSON_TYPE.OBJECT:
+            if (structValue is null)
+                structValue = new Struct;
+            structValue.fields = jsonValue.object.byKeyValue.map!(a => tuple(a.key, new Value(a.value))).assocArray;
+            break;
+        case JSON_TYPE.ARRAY:
+            if (listValue is null)
+                listValue = new ListValue;
+            listValue.values = jsonValue.array.map!(a => new Value(a)).array;
+            break;
+        case JSON_TYPE.TRUE:
+            boolValue = true;
+            break;
+        case JSON_TYPE.FALSE:
+            boolValue = false;
+            break;
+        default:
+            throw new ProtobufException("Unexpected JSON type");
+        }
+
+        return this;
+    }
+}
+
+class ListValue
+{
+    @Proto(1) Value[] values = defaultValue!(Value[]);
+
+    this()
+    {
+    }
+
+    this(JSONValue value)
+    {
+        fromJSONValue(value);
+    }
+
+    JSONValue toJSONValue()()
+    {
+        import std.array : array;
+        import std.algorithm : map;
+
+        return JSONValue(values.map!(a => a.toJSONValue).array);
+    }
+
+    auto fromJSONValue()(JSONValue value)
+    {
+        import std.array : array;
+        import std.algorithm : map;
+        import std.exception : enforce;
+        import std.json : JSON_TYPE;
+
+        enforce!ProtobufException(value.type == JSON_TYPE.ARRAY, "JSON array expected");
+        values = value.array.map!(a => new Value(a)).array;
+
+        return this;
+    }
+}
+
+enum NullValue
```

Comment:
How about using D's native `typeof(null)` type?

Reply:
This would require specialized implementation for the binary encoding, which uses a 0 integer. `NullValue` as `enum` reflects the definition in the `.proto` file which in turn reflects the binary encoding.